src/storage/buffer_pool_manager/buffer_pool.rs

```diff
@@ -0,0 +1,432 @@
+        assert!(entries_count > 0);
+        assert!(page_size > 0);
+
+        Self {
+            page_size,
+            buffer: UnsafeCell::new(vec![0x00u8; entries_count * page_size]),
+        }
+    }
+
+    /* private */
+    fn get_frame_range(&self, index: usize, offset: usize) -> Range<usize> {
+        (self.page_size * index + offset)..(self.page_size * (index + 1))
+    }
+}
+
+unsafe impl Send for FrameBuffer {}
+unsafe impl Sync for FrameBuffer {}
+
+#[derive(Debug)]
+pub struct BufferPoolImpl {
+    pub entries_count: usize,
+    buffer_map: DashMap<BufferPoolKey, usize>,
+    page_table: PageTable,
+    frame_buffer: FrameBuffer,
+}
+
+impl BufferPoolImpl {
+    pub fn new(entries_count: usize, page_size: usize) -> Self {
+        assert!(entries_count > 0);
+        assert!(page_size > 0);
+
+        Self {
+            entries_count,
+            buffer_map: DashMap::default(),
+            page_table: PageTable::new(entries_count),
+            frame_buffer: FrameBuffer::new(entries_count, page_size),
+        }
+    }
+}
+
+impl BufferPool for BufferPoolImpl {
+    fn get_page(&self, id: BufferPoolKey) -> PageReadLock {
+        if let Some(buffer_id) = self.buffer_map.get(&id) {
+            // Safety: `page_table` will never reallocate over the lifetime of the program.
+            let page_table = unsafe { self.page_table.table.get().as_ref().unwrap() };
+            let page_table_item_lock = page_table.get(*buffer_id).unwrap();
+
+            let page_table_item_read_lock_raw = unsafe { page_table_item_lock.raw() };
+            page_table_item_read_lock_raw.lock_shared();
+
+            // Safety: We have a read lock over `page_table_item_read_lock`.
+            let page_table_item_read_lock_data = page_table_item_lock.data_ptr();
+            let page_table_item_read_lock_data_ref =
+                unsafe { page_table_item_read_lock_data.as_ref().unwrap() };
+
+            // Safety: We have a read lock over `page_table_item_read_lock`.
+            let PageTableItem::Occupied { metadata, frame_lock } = page_table_item_read_lock_data_ref else { panic!("Not occupied!") };
+            let frame_read_lock_raw = unsafe { frame_lock.raw() };
+            frame_read_lock_raw.lock_shared();
+
+            // Safety: We have a read lock over `frame_read_lock`.
+            let buffer = unsafe { self.frame_buffer.buffer.get().as_ref().unwrap() };
+            let data = buffer
+                .get(self.frame_buffer.get_frame_range(*buffer_id, 0))
+                .unwrap();
+
+            {
+                let mut page_table_item_metadata_lock = metadata.lock();
+                page_table_item_metadata_lock.access_count += 1;
+            }
+
+            PageReadLock::new(page_table_item_read_lock_raw, frame_read_lock_raw, data)
+        } else {
+            // TODO: Not in the buffer pool.
+            todo!()
+        }
+    }
+
+    fn mutate_page(&self, id: BufferPoolKey, offset: usize, data: Vec<u8>) {
```

Comment:
Hm I'm not liking the idea of having to copy data... but maybe it's fine? Alternatively, thoughts on maybe passing a closure that takes in the data buffer as mutable to allow the caller to define what operations need to be done on the data buffer?